### PR TITLE
[DON'T MERGE] Go: fill more fields

### PR DESCRIPTION
**Pull request by masatake**

### CODE REVIEW

File: main/entry.c [hunk 1]
```diff
@@ -826,7 +840,22 @@ extern void getTagScopeInformation (tagEntryInfo *const tag,
...
-			*name = tag->extensionFields.scopeName;
+		{
+			if (tag->extensionFields.scopeIndex != CORK_NIL &&
+				(!isTagExtraBitMarked(tag, XTAG_QUALIFIED_TAGS)))
+			{
+				/* Though the tag has filled scopeName field ,
+				   it is a cache for building FQ tag.
+				   In the case, the filed filled with full qualified
+				   scope information. We will provide just one upper level
+				   scope information for this  no-qualified tag. */
```
[the remark this answers is not in the document]
Comment (masatake) [reply main/entry.c]: I see.

File: main/htable.h [hunk 2]
```diff
@@ -45,4 +47,11 @@ extern bool    hashTableDeleteItem  (hashTable *htable, void *key);
 extern void       hashTableForeachItem (hashTable *htable, hashTableForeachFunc proc, void *user_data);
 extern int        hashTableCountItem   (hashTable *htable);
 
+extern hashTable* hashTableIntNew (unsigned int size,
+								   hashTableHashFunc hashfn,
+								   hashTableEqualFunc equalfn,
+								   hashTableFreeFunc keyfreefn);
+#define HT_PTR_TO_INT(P) ((int)(intptr_t)(P))
+#define HT_INT_TO_PTR(P) ((void*)(intptr_t)(P))
```
Comment (masatake) [inline main/htable.h]: @b4n, do you think above two lines are acceptable?
Frequently, I need a data structure for mapping from a C string to cork index.

Comment (b4n) [reply main/htable.h]: @masatake yeah, looks good.

Comment (masatake) [reply main/htable.h]: Thank you.

File: main/entry.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (masatake) [reply main/entry.c]: I will withdraw this "[SELF-INCOMPATIBLE]main: ..." change.
As @b4n implied, the change is dirty. After getting the comment I have rewritten the code .
The new code is much clearer than original. However, I will not push the new code to here.

My original idea is that a non-fq tag shuold have a non-fq scope, and a fq tag should have a fq scope. My changes, old one and new one are implementing this idea. The idea itself is not so bad.
However, today, I find even C parser doesn't follow this idea:
```
[yamato@slave]~/var/ctags-github% cat /tmp/foo.c
struct foo {
  struct bar {
    int x;
  } y;
} z;
[yamato@slave]~/var/ctags-github% ./ctags -o - --extras=+q foo.c                        
ctags: Warning: cannot open input file "foo.c" : No such file or directory
[yamato@slave]~/var/ctags-github% ./ctags -o - --extras=+q /tmp/foo.c
bar	/tmp/foo.c	/^  struct bar {$/;"	s	struct:foo	file:
foo	/tmp/foo.c	/^struct foo {$/;"	s	file:
foo::bar	/tmp/foo.c	/^  struct bar {$/;"	s	struct:foo	file:
foo::bar::x	/tmp/foo.c	/^    int x;$/;"	m	struct:foo::bar	typeref:typename:int	file:
foo::y	/tmp/foo.c	/^  } y;$/;"	m	struct:foo	typeref:struct:foo::bar	file:
x	/tmp/foo.c	/^    int x;$/;"	m	struct:foo::bar	typeref:typename:int	file:
y	/tmp/foo.c	/^  } y;$/;"	m	struct:foo	typeref:struct:foo::bar	file:
z	/tmp/foo.c	/^} z;$/;"	v	typeref:struct:foo
```

See the scope infor for `y`. Here `y` is a non-fq tag entry. However, its scope is represented in fq style: "struct foo::bar".

Introducing my change may bring much more confusion to users.

If we will work on this item: whetehr we should use  fq or non-fq for scope field, I think we should introduce a new field scope2:.

For the parson really interesting this topic, I would like to paste my new code here.
```
commit 7c1197bb60da4dd7f48ef5888c8f2d98c4a7058f (revise-scope-field-for-autofq-fq-tag)
Author: Masatake YAMATO <yamato@redhat.com>
Date:   Sun Sep 16 04:12:46 2018 +0900

    [SELF-INCOMPATIBLE]main: don't use qualified name in scope field of non-qualified tag
    
    Signed-off-by: Masatake YAMATO <yamato@redhat.com>

diff --git a/main/entry.c b/main/entry.c
index b11d33c5..9c84e74b 100644
--- a/main/entry.c
+++ b/main/entry.c
@@ -748,6 +748,20 @@ extern void truncateTagLineAfterTag (
        }
 }
 
+static const char *getScopeNameFromCorkQueue (const tagEntryInfo * inner_scope)
+{
+       const tagEntryInfo *scope = inner_scope;
+
+       while (scope)
+       {
+               if (!scope->placeholder)
+                       return scope->name;
+               else
+                       scope = getEntryInCorkQueue (scope->extensionFields.scopeIndex);
+       }
+       return "";
+}
+
 static char* getFullQualifiedScopeNameFromCorkQueue (const tagEntryInfo * inner_scope)
 {
 
@@ -793,7 +807,7 @@ static char* getFullQualifiedScopeNameFromCorkQueue (const tagEntryInfo * inner_
 }
 
 extern void getTagScopeInformation (tagEntryInfo *const tag,
-                                   const char **kind, const char **name)
+                                   const char ** kind, const char ** name, bool fq)
 {
        if (kind)
                *kind = NULL;
@@ -806,16 +820,25 @@ extern void getTagScopeInformation (tagEntryInfo *const tag,
            && TagFile.corkQueue.count > 0)
        {
                const tagEntryInfo * scope;
-               char *full_qualified_scope_name;
 
                scope = getEntryInCorkQueue (tag->extensionFields.scopeIndex);
-               full_qualified_scope_name = getFullQualifiedScopeNameFromCorkQueue(scope);
-               Assert (full_qualified_scope_name);
-
-               /* Make the information reusable to generate full qualified entry, and xformat output*/
                tag->extensionFields.scopeLangType = scope->langType;
                tag->extensionFields.scopeKindIndex = scope->kindIndex;
-               tag->extensionFields.scopeName = full_qualified_scope_name;
+
+               /* Setting fq to true for
+                  making the information reusable to generate full qualified entry,
+                  and xformat output*/
+               if (fq)
+               {
+                       char *full_qualified_scope_name;
+                       full_qualified_scope_name = getFullQualifiedScopeNameFromCorkQueue(scope);
+                       Assert (full_qualified_scope_name);
+                       tag->extensionFields.scopeFQName = full_qualified_scope_name;
+               }
+               const char *scope_name;
+               scope_name = getScopeNameFromCorkQueue (scope);
+               Assert (scope_name);
+               tag->extensionFields.scopeName = eStrdup (scope_name);
        }
 
        if (tag->extensionFields.scopeKindIndex != KIND_GHOST_INDEX  &&
@@ -831,7 +854,12 @@ extern void getTagScopeInformation (tagEntryInfo *const tag,
                        *kind = kdef->name;
                }
                if (name)
-                       *name = tag->extensionFields.scopeName;
+               {
+                       if (fq && tag->extensionFields.scopeFQName)
+                               *name = tag->extensionFields.scopeFQName;
+                       else
+                               *name = tag->extensionFields.scopeName;
+               }
        }
 }
 
@@ -1029,6 +1057,8 @@ static void recordTagEntryInQueue (const tagEntryInfo *const tag, tagEntryInfo*
                slot->extensionFields.inheritance = eStrdup (slot->extensionFields.inheritance);
        if (slot->extensionFields.scopeName)
                slot->extensionFields.scopeName = eStrdup (slot->extensionFields.scopeName);
+       if (slot->extensionFields.scopeFQName)
+               slot->extensionFields.scopeFQName = eStrdup (slot->extensionFields.scopeFQName);
        if (slot->extensionFields.signature)
                slot->extensionFields.signature = eStrdup (slot->extensionFields.signature);
        if (slot->extensionFields.typeRef[0])
@@ -1100,6 +1130,8 @@ static void clearTagEntryInQueue (tagEntryInfo* slot)
                eFree ((char *)slot->extensionFields.inheritance);
        if (slot->extensionFields.scopeName)
                eFree ((char *)slot->extensionFields.scopeName);
+       if (slot->extensionFields.scopeFQName)
+               eFree ((char*)slot->extensionFields.scopeFQName);
        if (slot->extensionFields.signature)
                eFree ((char *)slot->extensionFields.signature);
        if (slot->extensionFields.typeRef[0])
@@ -1287,7 +1319,7 @@ extern void uncorkTagFile(void)
                if (doesInputLanguageRequestAutomaticFQTag ()
                    && isXtagEnabled (XTAG_QUALIFIED_TAGS)
                    && (tag->extensionFields.scopeKindIndex != KIND_GHOST_INDEX)
-                   && tag->extensionFields.scopeName
+                   && tag->extensionFields.scopeFQName
                    && tag->extensionFields.scopeIndex)
                        makeQualifiedTagEntry (tag);
        }
@@ -1393,9 +1425,15 @@ extern int makeQualifiedTagEntry (const tagEntryInfo *const e)
 
                fqn = vStringNewOrClearWithAutoRelease (fqn);
 
-               if (e->extensionFields.scopeName)
+               const char *scopeName = NULL;
+               if (e->extensionFields.scopeFQName)
+                       scopeName = e->extensionFields.scopeFQName;
+               else if (e->extensionFields.scopeName)
+                       scopeName = e->extensionFields.scopeName;
+
+               if (scopeName)
                {
-                       vStringCatS (fqn, e->extensionFields.scopeName);
+                       vStringCatS (fqn, scopeName);
                        xk = e->extensionFields.scopeKindIndex;
                        sep = scopeSeparatorFor (e->langType, e->kindIndex, xk);
                        vStringCatS (fqn, sep);
diff --git a/main/entry.h b/main/entry.h
index becd41de..7b149624 100644
--- a/main/entry.h
+++ b/main/entry.h
@@ -77,6 +77,9 @@ struct sTagEntryInfo {
                                             This field is meaningful if the value
                                             is not CORK_NIL and scope[0]  and scope[1] are
                                             NULL. */
+               const char* scopeFQName; /* This is not for a parser.
+                                                                       The main part use this for implementing
+                                                                       autoFQTag feature. */
 
                const char* signature;
```

@b4n, thank you very much for giving me a chance to rethink my change.

**Comment by masatake** [comment]
I withdraw this.
I opened #1974 and push reivised code to it.

**Pull request closed without merging**